[JSC] Optimize String#substring

Constellation opened the pull request:
#### 3de7e6fba239b0a69f618faa82c7cd642c57726b
<pre>
[JSC] Optimize String#substring
<a href="https://bugs.webkit.org/show_bug.cgi?id=244754">https://bugs.webkit.org/show_bug.cgi?id=244754</a>
&lt;rdar://99770218&gt;

Reviewed by Mark Lam.

This patch optimizes String#substring.

1. We avoid allocating the whole string when just comparing with substring. (JSString::equalSlowCase change).
   We get StringView with unsafeView, and use ensureStillAliveHere to keep both string cells alive.
2. We optimize JSString::equalSlowCase&apos;s equal function using StringView.
3. Optimize String#substring runtime function with int32 input fast path.
4. Add DFG StringSubstring node to handle String#substring efficiently in DFG / FTL. We also add StrengthReduction rule
   for String#substring so that we can constant-fold the result if arguments are the constants.

We observed improvements in microbenchmarks.
                                                    ToT                     Patched

    string-substring-constants                30.9691+-0.7392     ^      6.9364+-0.1119        ^ definitely 4.4647x faster
    string-substring-constants-binary         29.5540+-0.0829     ^     12.0035+-0.2692        ^ definitely 2.4621x faster
    string-starts-with-mod-prototype          69.7019+-0.6867     ^     56.8859+-0.4956        ^ definitely 1.2253x faster
    string-substring                          30.9168+-0.2336     ^     18.9088+-0.1950        ^ definitely 1.6350x faster
    string-substring-length-constant          30.6379+-0.2457     ^      6.9260+-0.1237        ^ definitely 4.4236x faster
    string-substring-constants-identity        9.5178+-0.1149     ^      2.9170+-0.0326        ^ definitely 3.2629x faster
    string-starts-with                         2.9473+-0.0220            2.9296+-0.0749
    string-starts-with-mod                    31.1585+-0.1608     ^     12.9144+-0.0646        ^ definitely 2.4127x faster

* JSTests/microbenchmarks/string-starts-with-mod-prototype.js: Added.
(shouldBe):
(String.prototype._startsWith):
(test1):
(test2):
* JSTests/microbenchmarks/string-starts-with-mod.js: Added.
(shouldBe):
(_startsWith):
(test1):
(test2):
* JSTests/microbenchmarks/string-starts-with.js: Added.
(shouldBe):
(test1):
(test2):
* JSTests/microbenchmarks/string-substring-constants-binary.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants-identity.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-length-constant.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring.js: Added.
(shouldBe):
(test1):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/RegExpPrototype.js:
(linkTimeConstant.getSubstitution):
(overriddenName.string_appeared_here.replace):
(overriddenName.string_appeared_here.split):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.repeatCharactersSlowPath):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileStringSubstring):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.cpp:
(JSC::intrinsicName):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::stringProtoSubstringFunction const):
* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSString::equalSlowCase const):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::stringSubstringImpl): Deleted.
* Source/JavaScriptCore/runtime/StringPrototype.h:
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::extractSubstringOffsets):
(JSC::stringSubstring):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::equalCommon):
* Source/WTF/wtf/text/StringView.h:
(WTF::equal):

Canonical link: <a href="https://commits.webkit.org/255030@main">https://commits.webkit.org/255030@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e78d8f150ed312a9c065bd38fed901f6a15e2687

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91050 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/35 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21580 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100475 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/159570 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/40 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83406 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97174 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96706 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/37 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27018 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70052 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82545 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35199 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15668 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77552 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32996 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16662 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26716 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3500 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36778 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80150 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38705 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35746 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17563 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -3685,6 +3685,24 @@ bool ByteCodeParser::handleIntrinsicCall(Node* callee, Operand result, Intrinsic
+            setResult(resultNode);
+            return true;
+        }
+
```
The code here is identical to the `StringPrototypeSliceIntrinsic` case except for the `NodeType` passed to `addToGraph`.  Why not just use the same code body and just parameterize the selection of the `NodeType`?

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -1409,6 +1409,11 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+        setTypeForNode(node, SpecString);
+        break;
+    }
+
```
This code is identical to the `StringSlice` case.  Why not just add `case StringSubstring:` to `case StringSlice:`?  You already group the `StringSubstring` case with the `StringSlice` case in many other places.

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -1234,25 +1233,29 @@ static EncodedJSValue stringSubstringImpl(JSGlobalObject* globalObject, CallFram
+
     int len = jsString->length();
-    RELEASE_ASSERT(len >= 0);
+    ASSERT(len >= 0);
```
Let's keep this as a `RELEASE_ASSERT`.

Constellation replied:
I think this is no longer meaningful since we are validating this length is within `[0, INT32_MAX]` range in JSString construction & String construction. Previously, we were using `UINT32_MAX` for String max length, and it caused problem here, but it never happens now. I'll keep it for now, but I'll upload a patch changing it to ASSERT.

MenloDorian commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -197,6 +197,7 @@ class StringView final {
     explicit StringView(const char*);
 
     friend bool equal(StringView, StringView);
+    friend bool equal(StringView, StringView, unsigned);
```
Please add the name `length` to the 3rd argument.  I think it adds meaningful info for someone who is only reading the header file.

Constellation replied:
Added.

MenloDorian reviewed (approved):
r=me with improvements.

Constellation committed:
[JSC] Optimize String#substring
https://bugs.webkit.org/show_bug.cgi?id=244754
<rdar://99770218>

Reviewed by Mark Lam.

This patch optimizes String#substring.

1. We avoid allocating the whole string when just comparing with substring. (JSString::equalSlowCase change).
   We get StringView with unsafeView, and use ensureStillAliveHere to keep both string cells alive.
2. We optimize JSString::equalSlowCase's equal function using StringView.
3. Optimize String#substring runtime function with int32 input fast path.
4. Add DFG StringSubstring node to handle String#substring efficiently in DFG / FTL. We also add StrengthReduction rule
   for String#substring so that we can constant-fold the result if arguments are the constants.

We observed improvements in microbenchmarks.
                                                    ToT                     Patched

    string-substring-constants                30.9691+-0.7392     ^      6.9364+-0.1119        ^ definitely 4.4647x faster
    string-substring-constants-binary         29.5540+-0.0829     ^     12.0035+-0.2692        ^ definitely 2.4621x faster
    string-starts-with-mod-prototype          69.7019+-0.6867     ^     56.8859+-0.4956        ^ definitely 1.2253x faster
    string-substring                          30.9168+-0.2336     ^     18.9088+-0.1950        ^ definitely 1.6350x faster
    string-substring-length-constant          30.6379+-0.2457     ^      6.9260+-0.1237        ^ definitely 4.4236x faster
    string-substring-constants-identity        9.5178+-0.1149     ^      2.9170+-0.0326        ^ definitely 3.2629x faster
    string-starts-with                         2.9473+-0.0220            2.9296+-0.0749
    string-starts-with-mod                    31.1585+-0.1608     ^     12.9144+-0.0646        ^ definitely 2.4127x faster

* JSTests/microbenchmarks/string-starts-with-mod-prototype.js: Added.
(shouldBe):
(String.prototype._startsWith):
(test1):
(test2):
* JSTests/microbenchmarks/string-starts-with-mod.js: Added.
(shouldBe):
(_startsWith):
(test1):
(test2):
* JSTests/microbenchmarks/string-starts-with.js: Added.
(shouldBe):
(test1):
(test2):
* JSTests/microbenchmarks/string-substring-constants-binary.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants-identity.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-length-constant.js: Added.
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring.js: Added.
(shouldBe):
(test1):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/RegExpPrototype.js:
(linkTimeConstant.getSubstitution):
(overriddenName.string_appeared_here.replace):
(overriddenName.string_appeared_here.split):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.repeatCharactersSlowPath):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileStringSubstring):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.cpp:
(JSC::intrinsicName):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::stringProtoSubstringFunction const):
* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSString::equalSlowCase const):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::stringSubstringImpl): Deleted.
* Source/JavaScriptCore/runtime/StringPrototype.h:
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::extractSubstringOffsets):
(JSC::stringSubstring):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::equalCommon):
* Source/WTF/wtf/text/StringView.h:
(WTF::equal):

Canonical link: https://commits.webkit.org/255030@main
